src/main.ts

```diff
@@ -1111,57 +1116,71 @@ function generateFromJson(ctx: Context, fullName: string, messageDesc: Descripto
         const fieldType = toTypeName(ctx, messageDesc, field);
         const i = maybeCastToNumber(ctx, messageDesc, field, 'key');
         chunks.push(code`
-          message.${fieldName} = Object.entries(object.${jsonName} ?? {}).reduce<${fieldType}>((acc, [key, value]) => {
-            acc[${i}] = ${readSnippet('value')};
-            return acc;
-          }, {});
+          ${fieldName}: ${ctx.utils.isObject}(object.${jsonName})
+            ? Object.entries(object.${jsonName}).reduce<${fieldType}>((acc, [key, value]) => {
+                acc[${i}] = ${readSnippet('value')};
+                return acc;
+              }, {})
+            : {},
         `);
       } else {
         const readValueSnippet = readSnippet('e');
         if (readValueSnippet.toString() === code`e`.toString()) {
-          chunks.push(code`message.${fieldName} = Array.isArray(object?.${jsonName}) ? [...object.${jsonName}] : [];`);
+          chunks.push(code`${fieldName}: Array.isArray(object?.${jsonName}) ? [...object.${jsonName}] : [],`);
         } else {
           // Explicit `any` type required to make TS with noImplicitAny happy. `object` is also `any` here.
           chunks.push(code`
-            message.${fieldName} = Array.isArray(object?.${jsonName}) ? object.${jsonName}.map((e: any) => ${readValueSnippet}): [];
+            ${fieldName}: Array.isArray(object?.${jsonName}) ? object.${jsonName}.map((e: any) => ${readValueSnippet}): [],
           `);
         }
       }
     } else if (isWithinOneOfThatShouldBeUnion(options, field)) {
-      chunks.push(code`if (${ctx.utils.isSet}(object.${jsonName})) {`);
+      const oneofFields = messageDesc.field
+        .filter(isWithinOneOf)
+        .filter((otherField) => otherField.oneofIndex === field.oneofIndex);
+
       const oneofName = maybeSnakeToCamel(messageDesc.oneofDecl[field.oneofIndex].name, options);
-      chunks.push(code`
-        message.${oneofName} = { $case: '${fieldName}', ${fieldName}: ${readSnippet(`object.${jsonName}`)} }
-      `);
-      chunks.push(code`}`);
+
+      // First case, output field name
```

Comment:
Ah nice, I see...

Seems like this is relying on all of the oneofFields being grouped together, sequentially in the incoming `messageDesc.field` list.

I wouldn't be surprised if this is only true for our relatively clean `*.proto` test files, and that real-world proto files this may not be the case? Maybe? Not sure.

Actually I guess if the order is based on `*.proto` order than we should be good, so maybe this is just fine. I'm good keeping it as it is, b/c it's simpler this way, and we can see if anyone ends up reporting not-actually-in-order oneof fields as users try it out.

Reply:
Yeah, this one got me thinking for some time as well...

As far as I can reason through this, it shouldn't matter in what order the fields are delivered, and even whether there are random unrelated fields in between.

We are generating the fields in order of `messageDesc.field`, and the index checks (first / last) are based on a subset of the same sequence. So no matter how the order might be messed up, the first will always be the first etc. If that makes sense.

Reply:
I realize that the above merely guarantees correct detection of the _first_ and _last_ item, but if indeed _somehow_ oneof field members were interspersed with other random fields, we would mess up the output.

However, I do wish to rely on the protobuf sdk to keep the fields in the declared order, and in that case, it should all work out.

I've changed this part of the code a bit to be more readable, hope that helps.